### ISSUE TITLE
Writing To A Global Variable When Tests Are Run In Parallel

### ISSUE THREAD

**Issue by bhreinb** [cucumber/cucumber-js#674]
Hi There

I'm not sure whether this is the right thread to post this comment in but I was hoping somebody can help with this?

I have the following configuration data snippet

```
var _Global;

beforeLaunch: function()
{

    var Fsp = require('fs-promise'), mkdirp = require('mkdirp'), ResultsDirectory = './Results';

    //Minimum MetaData That Will Get Wrote To Report...
    global.MetaData =
    {

        OsPlatform: null,

        AngularVersion: null

    };

    _Global = global;

    return Fsp.exists(ResultsDirectory)
    .then(function(exists) { if(exists) { return Fsp.remove(ResultsDirectory); } })
    .then(function()
    {

        return mkdirp(ResultsDirectory, function (err)
        {

            if (err)

                console.error(err);

            else

                console.log('Directory To Store Test Results Created As Expected!!!');

        });

    });

},

onPrepare: function()
{

    var _this = this;

    browser.ignoreSynchronization = false;

    browser.getProcessedConfig().then(function(currentConfig)
    {

        console.log('For This Test We Are Using Browser: ' + currentConfig.capabilities.browserName);

        console.log('Full Details Listed Below:');

        console.log(JSON.stringify(currentConfig.capabilities, null, 4));

    });

    return browser.get(browser.baseUrl)
    .then(function() {return browser.executeScript("return window.angular.version.full;")})
    .then(function(AngularVersion)
    {

        _Global.MetaData.OsPlatform = require('os').platform().toUpperCase();

        //BrowserName & Version...
        browser.getCapabilities().then(function (Caps)
        {

            var BrowserObject = JSON.parse("{\"" + Caps.get('browserName') + "\": \"" +Caps.get('version')+ "\"}");

            return Object.assign(_Global.MetaData, BrowserObject);

        });

        return _Global.MetaData.AngularVersion = AngularVersion;

    });


},

afterLaunch: function(ExitCode)
{

    console.log('Finished With Exit Code ' +ExitCode);

    return Reports.GetAllJsonFilesInResultsDirectory()
    .then(function() {return Reports.MergeJsonFilesToOneObject();})
    .then(function() {return Reports.WriteJsonStringToFileSystem();})
    .then(function() {return Reports.CreateJunitXmlReport();})
    .then(function() {return Reports.CreateHtmlReports();})
    .then(function(Results)
    {

        Results.forEach(function(Result)
        {

            console.log(Result);

        });

    });

}

```
When I run the tests on a single browser instance this works every time but when I run in parallel this fails on function "onPrepare", namely it throws an exception. Does "cucumber.js" have a global space that test data could be wrote to or any similar equivalent?

Basically the use here is to write the browser name & version to a global variable "MetaData" which will get wrote to the final report generated. So in parallel tests you would have browser information for ie, firefox & chrome wrote to object "MetaData". Many thanks in advance as to any guidance anyone provides.

**Comment by bhreinb** [comment]
This can be closed as I found a solution around this.

**Comment by danielcb29** [comment]
Hi @bhreinb , can you share the solution you found?

**Comment by bhreinb** [comment]
Hi @danielcb29 

This question definitely shouldn't exist in this space. This is a protractor life cycle question more so than a cucumberjs question. In any case my solution is detailed below

My use was that I wanted to collect metadata across different browsers and merge that to the one object at the end of a test run. Generally the metadata was more or less the same per browser instance the only difference been was the browser name and version. Solving this required understanding of the protractor life cycle

```
beforeLaunch ~ is executed once only if single instance browser testing or parallel testing
onPrepare ~ is executed on each browser the test is run on 
afterLaunch  ~ is executed once only if single instance browser testing or parallel testing
```

based on that info I collected all the metadata in the onPrepare block and wrote to a JSON file at a know location (I included the browser name & version when writing this). In the afterlaunch life cycle event i collected all the JSON files using regular expression matching and merged them, thus getting me the one JSON file that .

Unfortunately you can't write each browser object to the global space because when we launch in parallel it launches the tests under different PIDs hence the inability to write this in the global space. Hope this helps.

**Comment by danielcb29** [comment]
Thanks @bhreinb for share it! 👍 , my problem was pretty simple I solved it using the global scope. I wanted to store a server connection for all scenarios.
